fix(ast/estree): order fields same as Acorn

overlookmotel opened the pull request:
Use the `#[estree(field_order(...))]` attribute introduced in #9127 to re-order fields in the ESTree AST so they match Acorn exactly.

Make the ESTree conformance tester stricter - it now errors on incorrect field order.

All conformance tests still passing after this PR.

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9128?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#9128** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9128?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9128?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#9127** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9127?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9125** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9125?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9124** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9124?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9118** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9118?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9117** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9117?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9116** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9116?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9128).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Boshen commented:
### Merge activity

* **Feb 14, 8:03 PM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
fix(ast/estree): order fields same as Acorn (#9128)

Use the `#[estree(field_order(...))]` attribute introduced in #9127 to re-order fields in the ESTree AST so they match Acorn exactly.

Make the ESTree conformance tester stricter - it now errors on incorrect field order.

All conformance tests still passing after this PR.

knightedcodemonkey commented:
I was caught off guard by `body: FunctionBody | null;` for the `Function` interface. Curious why it is ever `null` even after narrowing the node type to `FunctionExpression`.

camc314 commented:
> I was caught off guard by `body: FunctionBody | null;` for the `Function` interface. Curious why it is ever `null` even after narrowing the node type to `FunctionExpression`.

In TS, when there's a function overload e.g.

```ts
function foo(): void
function foo(arg0?: string): void { }
```

the overload does not have a function body.

knightedcodemonkey commented:
@camc314 thanks for that example. So in that case the `TSDeclareFunction` which is part of the `FunctionType` definition is causing a `body: null` situation (first line in example).

I guess there's another scenario where narrowing the function `type` to `FunctionExpression` would cause a `null` body? Maybe something with `TSEmptyBodyFunctionExpression`?

I wonder if it might be beneficial to create distinct types for the TS specific function types and the pure JS function types so that narrowing against specific function types doesn't require further narrowing on the presence of a `body`? For example, I'd like to avoid a situation like:

```js
switch (func.type) {
  case 'FunctionExpression':
    func.body.start // <--- 'func.body' is possibly `null` TS error
```

camc314 commented:
Interestingly `FunctionExpression`, should always have a body defined (missing bodies are only allowed for function declarations), but i guess the types don't reflect that.

https://github.com/oxc-project/oxc/blob/8a3bba83569413a0deb9ea81a8559522554fead1/crates/oxc_ast/src/ast/js.rs#L1735-L1747

this is definitely in @overlookmotel terrority, but...

```ts
type Expression = Function // and some other items
```

Perhaps here it should be `type Expression = FunctionExpression /* other varients */`  to force the body to be defined?

overlookmotel commented:
@knightedcodemonkey Could you please open a separate issue for this (and include details if conversation with cam here)? I don't know the solution at present, and don't want this to get lost, which it surely will do being tacked on the end of an old PR.

knightedcodemonkey commented:
@overlookmotel good point, better to have this as a separate issue: https://github.com/oxc-project/oxc/issues/10988
